Keep @modifiers when parsing locales

madduck opened the pull request:
Locale modifiers ("@variants") are described in the GNU gettext documentation like this:

> The ‘@variant’ can denote any kind of characteristics that is not
> already implied by the language ll and the country CC. […] It can also
> denote a dialect of the language, …

Wherein Babel previously would discard these, this patch stores the modifier information in the `Locale` objects, handling string representation accordingly.

Not implemented is the lookup of a meaningful description of modifiers, but instead — for now — an identity mapping is provided.

All tests pass. Furthermore, integration with `flask-babel` has been verified to work.

Resolves: #946
Signed-off-by: martin f. krafft <madduck@madduck.net>

akx commented on babel/core.py:
```diff
@@ -191,10 +193,13 @@ def __init__(
 
         identifier = str(self)
-        if not localedata.exists(identifier):
+        withoutmodifier = identifier.split('@', 1)[0]
```
Can you move this to a property, e.g. `identifier_without_modifier`? I think we'd need to use this in the `_data` property too.

When you do that, can you also make it use `partition` instead:

```suggestion
        without_modifier = identifier.partition('@')[0]
```

madduck replied:
I can make this a property, sure, but I didn't, because `identifier` also isn't one.

I am unsure about `_data`. I didn't really dive this far on purpose because I didn't invest time trying to figure out where all the `*.dat` files are being created from.

As pertaining to the use of `partition`: I learnt something new. In this case, however, I don't think it matters much, but I'll happily change it.

akx commented on babel/core.py:
```diff
@@ -289,6 +294,11 @@ def parse(
         >>> Locale.parse('und_AT')
         Locale('de', territory='AT')
 
+        Modifiers are optional, and always at the end, separated by "@":
```
Might be clearer as
```suggestion
        An optional modifier can be added at the end, separated by "@".
```

akx commented on babel/core.py:
```diff
@@ -562,6 +582,24 @@ def variants(self) -> localedata.LocaleDataDict:
         """
         return self._data['variants']
 
+    @property
+    def modifiers(self) -> localedata.LocaleDataDict:
+        """Identity mapping of modifiers with "@" prefixed (Temporary implementation)
+
+        TODO: This is not yet implemented, as it would need modification of the
+        locale_data files, so instead it just returns the key for now, with "@"
+        prefixed.
+
+        >>> Locale('de', 'DE').modifiers['euro']
+        u'@euro'
+        """
+
+        class IdentityDict(localedata.LocaleDataDict):
+            def __getitem__(self, key: str) -> Any:
+                return f'@{key!s}'
+
+        return IdentityDict(self._data)
+
```
I'm not sure this needs to be implemented at all yet?

madduck replied:
Yeah, neither. I figure it doesn't hurt and provides a bit of an anchor to keep working on this, but the truth is that we probably will need to either remove it, or tap into the creation of `*.dat` files.

akx replied:
Let's go with removing it for now, please? That way we can come up with a proper API later and aren't stuck freezing this as a public API :)

akx commented on babel/core.py:
```diff
@@ -1112,26 +1150,32 @@ def negotiate_locale(preferred: Iterable[str], available: Iterable[str], sep: st
 
 
-def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str | None, str | None]:
+def parse_locale(identifier: str, sep: str = '_') \
+        -> tuple[str, str | None, str | None, str | None, str | None]:
```
```suggestion
def parse_locale(
    identifier: str,
    sep: str = '_',
) -> tuple[str, str | None, str | None, str | None, str | None]:
```

madduck replied:
No discussion. I was running `black` on the code, and it would have done more or less the same, but it also changed 350 other lines, so I hit 'undo' ;)

akx commented on babel/core.py:
```diff
@@ -1112,26 +1150,32 @@ def negotiate_locale(preferred: Iterable[str], available: Iterable[str], sep: st
+    ('en', 'US', None, 'POSIX', None)
+    >>> parse_locale('it_IT@euro')
+    ('it', 'IT', None, None, 'euro')
+    >>> parse_locale('it_IT@custom')
+    ('it', 'IT', None, None, 'custom')
```
I'd like to be sure this parses an empty modifier as None:
```suggestion
    >>> parse_locale('it_IT@custom')
    ('it', 'IT', None, None, 'custom')
    >>> parse_locale('it_IT@')
    ('it', 'IT', None, None, None)
```

madduck replied:
Makes sense! Thanks.

akx commented on babel/core.py:
```diff
@@ -1112,26 +1150,32 @@ def negotiate_locale(preferred: Iterable[str], available: Iterable[str], sep: st
 
     The default component separator is "_", but a different separator can be
-    specified using the `sep` parameter:
+    specified using the `sep` parameter. Note that an optional modifier is
+    always appended and separated with "@":
```
```suggestion
    specified using the `sep` parameter.
    
    The optional modifier component is always separated with "@":
```

madduck replied:
Done.

akx commented on babel/core.py:
```diff
@@ -1112,26 +1150,32 @@ def negotiate_locale(preferred: Iterable[str], available: Iterable[str], sep: st
 
     >>> parse_locale('zh-CN', sep='-')
-    ('zh', 'CN', None, None)
+    ('zh', 'CN', None, None, None)
```
```suggestion
    ('zh', 'CN', None, None, None)
    >>> parse_locale('zh-CN@custom', sep='-')
    ('zh', 'CN', None, None, 'custom')
```

madduck replied:
Cool.

akx commented on babel/core.py:
```diff
@@ -1158,13 +1201,13 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
     :raise `ValueError`: if the string does not appear to be a valid locale
                          identifier
     """
+    modifier = None
+    if '@' in identifier:
+        identifier, modifier = identifier.split('@', 1)
```
You can use `partition` here too.
```suggestion
    identifier, _, modifier = identifier.partition('@')
```

madduck replied:
Sweet! I learnt something.

akx commented on babel/core.py:
```diff
@@ -1190,22 +1233,23 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
         raise ValueError(f"{identifier!r} is not a valid locale identifier")
 
-    return lang, territory, script, variant
+    return lang, territory, script, variant, modifier
```
If you use `partition`:
```suggestion
    return lang, territory, script, variant, (modifier or None)
```

madduck replied:
Yeah, related to the change above. Makes tests pass.

akx commented on babel/core.py:
```diff
@@ -1190,22 +1233,23 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
 
 
-def get_locale_identifier(tup: tuple[str, str | None, str | None, str | None], sep: str = '_') -> str:
+def get_locale_identifier(tup: tuple[str, str | None, str | None, str | None, str | None], sep: str = '_') -> str:
```
This function should also support 4-tuples. In fact, it would be best if it supported any tuple size from 1 to 5. (I think the code already does, but the signature doesn't reflect that.)

madduck replied:
Hm, I am not aware of a way to represent this in the function signature. Yes, I think the code does support it already, but I am unsure how else you want the signature to look. Any further hints?

akx replied:
I think it'll be a bit unwieldy, but...

```suggestion
def get_locale_identifier(
    tup: tuple[str]
    | tuple[str, str | None]
    | tuple[str, str | None, str | None]
    | tuple[str, str | None, str | None, str | None]
    | tuple[str, str | None, str | None, str | None, str | None],
    sep: str = "_",
) -> str:
```

akx commented on babel/core.py:
```diff
@@ -1190,22 +1233,23 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
-    lang, territory, script, variant = tup + (None,) * (4 - len(tup))
-    return sep.join(filter(None, (lang, script, territory, variant)))
+    tup = tuple(tup[:5])
+    lang, territory, script, variant, modifier = tup + (None,) * (4 - len(tup))
```
I think the `4` here is off by one?

```suggestion
    lang, territory, script, variant, modifier = tup + (None,) * (5 - len(tup))
```

That said, that's quite the horrible expression...

madduck replied:
Good catch!

I cannot really think of a better way, other than to replace the tuple with e.g. a `collections.namedtuple` or otherwise class of its own.

akx commented on babel/core.py:
```diff
@@ -180,6 +181,7 @@ def __init__(
         :param territory: the territory (country or region) code
         :param script: the script code
         :param variant: the variant code
+        :param modifier: a modifier ('@variant')
```
Does this imply the modifier has the `@` intact here? It shouldn't, IMO..?

madduck replied:
I agree. I was studying the docs, and often saw '@variant', which I thought would be read as "at-variant", and so I included it. But it should probably go. Done.

akx reviewed (changes requested):
Thank you for the contribution!
Some initial thoughts and comments within; you'll probably also need to rebase since #949 just landed...

madduck commented:
@akx Thank you for your time. I addressed most of your comments, and left an open question in the above. Happy to proceed any way, just let me know what you prefer.

akx commented on babel/core.py:
```diff
@@ -1193,22 +1240,23 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
 
-    >>> get_locale_identifier(('de', 'DE', None, '1999'))
-    'de_DE_1999'
+    >>> get_locale_identifier(('de', 'DE', None, '1999', 'custom'))
+    'de_DE_1999@custom'
```
Can you make sure this works with elided bits in the middle too?
```suggestion
    >>> get_locale_identifier(('de', 'DE', None, '1999', 'custom'))
    'de_DE_1999@custom'
    >>> get_locale_identifier(('fi', None, None, None, 'custom'))
    'fi@custom'
```

akx commented on babel/core.py:
```diff
@@ -1193,22 +1240,23 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
     if parts:
         raise ValueError(f"{identifier!r} is not a valid locale identifier")
 
-    return lang, territory, script, variant
+    return lang, territory, script, variant, (modifier or None)
```
Actually, can you change this function to only return a 5-tuple if there is a modifier, and then add a `# TODO(3.0): always return a 5-tuple` sort of comment? We're trying pretty hard to not change public APIs.

akx commented on babel/core.py:
```diff
@@ -1193,22 +1222,37 @@ def parse_locale(identifier: str, sep: str = '_') -> tuple[str, str | None, str
-
-
-def get_locale_identifier(tup: tuple[str, str | None, str | None, str | None], sep: str = '_') -> str:
+    # TODO(3.0): always return a 5-tuple
+    if modifier:
+        return lang, territory, script, variant, modifier
+    else:
+        return lang, territory, script, variant
```
No need for the `else` :)
```suggestion
    # TODO(3.0): always return a 5-tuple
    if modifier:
        return lang, territory, script, variant, modifier
    return lang, territory, script, variant
```

madduck replied:
True, but I personally find it clearer that way. But happy to go without if you insist.

akx reviewed (changes requested):
Tests are failing, can you take a look? Thanks.

madduck commented:
Sorry for the failing tests, was due to the 4/5 tuple return depending on `modifier==None`. I am not actually very sure it's a good idea, but I see your point. Fixed now.

akx reviewed (approved):
Thanks, looks good!
